akka-projection-core/src/main/scala/akka/projection/scaladsl/SourceProvider.scala

```diff
@@ -26,6 +27,46 @@ trait SourceProvider[Offset, Envelope] {
 
 }
 
+/**
+ * By default, a `SourceProvider` uses the stored offset when starting the Projection. This offset can be adjusted
+ * by defining the `startOffset` function.
+ */
+trait CustomStartOffsetSourceProvider[Offset, Envelope] extends SourceProvider[Offset, Envelope] {
+
+  /**
+   * Adjust the start offset by defining a function that returns the offset to use.
+   *
+   * @param loadFromOffsetStore whether the offset should be loaded from the offset store or not, loaded offset
+   *                            is passed as the parameter to the `startOffset` function
+   * @param startOffset function from loaded offset (if any) to the adjusted offset
+   */
+  def withStartOffset(loadFromOffsetStore: Boolean, startOffset: Option[Offset] => Future[Option[Offset]]): Unit
+
+  def loadFromOffsetStore: Boolean
+
+  def startOffset: Option[Offset] => Future[Option[Offset]]
```

Comment:
those two are used by us in the Projection impl, but need to define them as public because SourceProvider can be implemented by user